### PR TITLE
[Fuchsia] Run tests with multiple packages

### PR DESCRIPTION
**Pull request by zijiehe-google-com**
This change enables tests requiring multiple packages in both debug and release builders; though the AOT test cannot be executed.
Also 4 out of 5 suites are excluded from release builder since some of the tests are failing; it's worth extra investigation.
MouseInputTest in theory is enabled, it indeed contains only disabled tests.

Bug: https://github.com/flutter/flutter/issues/140179

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: testing/fuchsia/run_tests.py [hunk 1]
```diff
@@ -102,18 +104,31 @@ def bundled_test_runner_of(target_id: str) -> BundledTestRunner:
...
+        os.remove(package)
+      except FileNotFoundError:
+        pass
+      os.symlink(original_package, package)
+      resolved_packages.append(package)
+    else:
+      resolved_packages.append(os.path.join(OUT_DIR, package))
   return BundledTestRunner(
-      target_id, [test['package'] for test in tests],
+      target_id, list(dict.fromkeys(resolved_packages)),
```
Comment (jrwang) [inline testing/fuchsia/run_tests.py:131]: Why is it necessary here and elsewhere to have the collection of packages be a list -> dict -> list again?

If we're worried about duplicates, it should be a set; if duplicates aren't possible, then we should just keep it a list.

File: testing/fuchsia/run_tests.py [hunk 2]
```diff
@@ -102,18 +104,31 @@ def bundled_test_runner_of(target_id: str) -> BundledTestRunner:
...
+    if 'package' in test:
+      packages.append(test['package'])
+    else:
+      assert 'packages' in test, \
+             'Expect either one package or a list of packages'
+      packages.extend(test['packages'])
+  resolved_packages = []
+  for package in list(dict.fromkeys(packages)):
+    if package.endswith('-0.far'):
+      # Make a symbolic link to match the name of the package itself without the
```
Comment (jrwang) [inline testing/fuchsia/run_tests.py:118]: Why is this necessary?

File: testing/fuchsia/run_tests.py (hunk 1 above)
Comment (jrwang) [inline testing/fuchsia/run_tests.py:131]: Why is it necessary here and elsewhere to have the collection of packages be a list -> dict -> list again?

If we're worried about duplicates, it should be a set; if duplicates aren't possible, then we should just keep it a list.

File: testing/fuchsia/run_tests.py [hunk 3]
```diff
@@ -82,18 +82,20 @@ def bundled_test_runner_of(target_id: str) -> BundledTestRunner:
...
               'test_command'].endswith('.cm'), tests
       )
   )
+  # TODO(zijiehe-google-com): Run tests with dart aot,
+  # https://github.com/flutter/flutter/issues/140179.
   tests = list(
       filter(
-          lambda test: 'package' in test and test['package'].endswith('-0.far'),
-          tests
+          lambda test: not 'run_with_dart_aot' in test or test[
```
Comment (jrwang) [inline testing/fuchsia/run_tests.py:97]: style: 'run_with_dart_aot' not in test

**Review by jrwang — CHANGES_REQUESTED** [review]

File: testing/fuchsia/run_tests.py (hunk 2 above)
Comment (zijiehe-google-com) [reply testing/fuchsia/run_tests.py:118]: In chromium test script, we do not expect the far file name to be renamed to `-0.far` (https://github.com/flutter/engine/blob/a7183373eb5a86ab85852b7a5b077fc73cb4f37f/tools/fuchsia/gen_package.py#L125).
I can definitely add a new parameter in the test script, but it seems not very necessary, using a symbolic link can work it around.

File: testing/fuchsia/run_tests.py (hunk 1 above)
Comment (zijiehe-google-com) [reply testing/fuchsia/run_tests.py:131]: Yes, changed to set.

File: testing/fuchsia/run_tests.py (hunk 3 above)
Comment (zijiehe-google-com) [reply testing/fuchsia/run_tests.py]: Done.

File: testing/fuchsia/run_tests.py [hunk 4]
```diff
@@ -102,18 +104,31 @@ def bundled_test_runner_of(target_id: str) -> BundledTestRunner:
...
+      original_package = package
+      package = os.path.join(OUT_DIR, package.replace('-0.far', '.far'))
+      try:
+        os.remove(package)
+      except FileNotFoundError:
+        pass
+      os.symlink(original_package, package)
+      resolved_packages.append(package)
+    else:
+      resolved_packages.append(os.path.join(OUT_DIR, package))
```
Comment (jrwang) [inline testing/fuchsia/run_tests.py:129]: This logic would be clearer with some different variable names:

```
if package.endswith('-0.far'):
   new_package_path = os.path.join(OUT_DIR, package.replace('-0.far', '.far'))
   try:
      os.remove(new_package_path)
   except FileNotFoundError:
      pass
   os.symlink(package_path, new_package_path)
   resolved_packages.add(new_package_path)
else:
   resolved_packages.add(os.path.join(OUT_DIR, package))
```

Comment (jrwang) [reply testing/fuchsia/run_tests.py:129]: Also it's not immediately clear to me why we need to remove the new_package_name file (in case it already exists? Is it possible for it to ever exist?) Maybe a comment would be useful here.

File: testing/fuchsia/run_tests.py [hunk 5]
```diff
@@ -102,18 +104,31 @@ def bundled_test_runner_of(target_id: str) -> BundledTestRunner:
           tests
```
Comment (jrwang) [inline testing/fuchsia/run_tests.py:104]: nit: 'variant' not in test (in previous line)

**Review by jrwang — APPROVED** [review]
LGTM (with changes)

File: testing/fuchsia/run_tests.py [hunk 6]
```diff
@@ -102,18 +104,31 @@ def bundled_test_runner_of(target_id: str) -> BundledTestRunner:
...
       )
   )
+  packages = []
```
[the remark this answers is not in the document]
Comment (zijiehe-google-com) [reply testing/fuchsia/run_tests.py:107]: Whoo, I failed into the trap of a List and couldn't escape from it anymore :|

Done.

File: testing/fuchsia/run_tests.py (hunk 4 above)
Comment (zijiehe-google-com) [reply testing/fuchsia/run_tests.py:129]: Yes, if running locally, the symbolic link won't be removed automatically. Added comment.

File: testing/fuchsia/run_tests.py (hunk 5 above)
Comment (zijiehe-google-com) [reply testing/fuchsia/run_tests.py:104]: Done.

**Comment by zijiehe-google-com** [comment]
Thank you.
